### PR TITLE
Fix classifier parsing in python 3.8 / python 3.9

### PR DESCRIPTION
**Pull request by a1lu**
The classifier parser did not work as expected on my machine.

In python3.8 and python3.9 the type `email.message import Message` is used as metadata type and the metadata are stored in `i_metadata._headers` as a list of tuples. To iterate over the classifiers in this list of tuples 
```
for classifier in i_metadata['Classifier']:
```
was used. Though this loop iterates only over the first classifiers characters and not over all classifiers, so the classifiers were never parsed.

I propose in this commit to generate a list of classifiers for python3.8/python3.9 and python3.10 separately and loop over the list of classifiers.

Please note:
I created this PR on github web, I guess the commit message needs to be adjusted before merging!

### CODE REVIEW

**Comment by jkowalleck** [comment]
Could you add proper (unit) tests to prove that your changes actually work? 
The tests will be running on CI with every supported python version on every operating system.


Additionally, could you please sign your commit according to <https://github.com/CycloneDX/cyclonedx-python/pull/431/checks?check_run_id=9224681013> ?

File: cyclonedx_py/parser/environment.py [hunk 1]
```diff
@@ -75,18 +75,23 @@ def __init__(self, use_purl_bom_ref: bool = False) -> None:
...
-                        )))
-                    elif str(classifier).startswith('License :: '):
-                        c.licenses.add(LicenseChoice(license_=License(
-                            license_name=str(classifier).replace('License :: ', '').strip()
-                        )))
+            if sys.version_info >= (3, 8):
+               if sys.version_info >= (3, 10):
+                   classifiers = i_metadata.get('Classifier', [])
+               else:
+                   classifiers = [value for (key, value) in i_metadata._headers if key =='Classifier']
```
Comment (jkowalleck) [inline cyclonedx_py/parser/environment.py:82]: :x: 
if non of these is true, then `classifiers` woyuld be undefined.
but it is treated as an iterable

Comment (jkowalleck) [inline cyclonedx_py/parser/environment.py:82]: :x: 
```py
classifiers = [value for (key, value) in i_metadata._headers if key =='Classifier']    
```
what is this supposed to do? could you describe what you think this does?

**Review by jkowalleck — CHANGES_REQUESTED** [review]

**Comment by jkowalleck** [comment]
thank you for the pull request, @a1lu .

please follow https://github.com/CycloneDX/cyclonedx-python/blob/master/CONTRIBUTING.md

File: cyclonedx_py/parser/environment.py (hunk 1 above)
Comment (a1lu) [reply cyclonedx_py/parser/environment.py:82]: On my machine (ubuntu, python 3.8.10), on a breakpoint in this line:
started command: /home/.../cyclonedx-python/venv/bin/python  /home/.../cyclonedx-python/cyclonedx_py/client.py -e 
```
>>> i
zipp 3.10.0 (/home/.../cyclonedx-python/venv/lib/python3.8/site-packages)

>>> type(i_metadata)
<class 'email.message.Message'>

>>> i_metadata.__dict__
{'policy': Compat32(), '_headers': [('Metadata-Version', '2.1'), ('Name', 'zipp'), ('Version', '3.10.0'), ('Summary', 'Backport of pathlib-compatible object wrapper for zip files'), ('Home-page', 'https://github.com/jaraco/zipp'), ('Author', 'Jason R. Coombs'), ('Author-email', 'jaraco@jaraco.com'), ('Classifier', 'Development Status :: 5 - Production/Stable'), ('Classifier', 'Intended Audience :: Developers'), ('Classifier', 'License :: OSI Approved :: MIT License'), ('Classifier', 'Programming Language :: Python :: 3'), ('Classifier', 'Programming Language :: Python :: 3 :: Only'), ('Requires-Python', '>=3.7'), ('License-File', 'LICENSE'), ('Provides-Extra', 'docs'), ('Requires-Dist', "sphinx (>=3.5) ; extra == 'docs'"), ('Requires-Dist', "jaraco.packaging (>=9) ; extra == 'docs'"), ('Requires-Dist', "rst.linker (>=1.9) ; extra == 'docs'"), ('Requires-Dist', "furo ; extra == 'docs'"), ('Requires-Dist', "jaraco.tidelift (>=1.4) ; extra == 'docs'"), ('Provides-Extra', 'testing'), ('Requires-Dist', "pytest (>=6) ; extra == 'testing'"), ('Requires-Dist', "pytest-checkdocs (>=2.4) ; extra == 'testing'"), ('Requires-Dist', "pytest-flake8 ; extra == 'testing'"), ('Requires-Dist', "flake8 (<5) ; extra == 'testing'"), ('Requires-Dist', "pytest-cov ; extra == 'testing'"), ('Requires-Dist', "pytest-enabler (>=1.3) ; extra == 'testing'"), ('Requires-Dist', "jaraco.itertools ; extra == 'testing'"), ('Requires-Dist', "func-timeout ; extra == 'testing'"), ('Requires-Dist', "jaraco.functools ; extra == 'testing'"), ('Requires-Dist', "more-itertools ; extra == 'testing'"), ('Requires-Dist', 'pytest-black (>=0.3.7) ; (platform_python_implementation != "PyPy") and extra == \'testing\''), ('Requires-Dist', 'pytest-mypy (>=0.9.1) ; (platform_python_implementation != "PyPy") and extra == \'testing\'')], '_unixfrom': None, '_payload': '.. image:: https://img.shields.io/pypi/v/zipp.svg\n   :target: `PyPI link`_\n\n.. image:: https://img.shields.io/pypi/pyversions/zipp.svg\n   :target: `PyPI link`_\n\n.. _PyPI link: https://pypi.org/project/zipp\n\n.. image:: https://github.com/jaraco/zipp/workflows/tests/badge.svg\n   :target: https://github.com/jaraco/zipp/actions?query=workflow%3A%22tests%22\n   :alt: tests\n\n.. image:: https://img.shields.io/badge/code%20style-black-000000.svg\n   :target: https://github.com/psf/black\n   :alt: Code style: Black\n\n.. .. image:: https://readthedocs.org/projects/skeleton/badge/?version=latest\n..    :target: https://skeleton.readthedocs.io/en/latest/?badge=latest\n\n.. image:: https://img.shields.io/badge/skeleton-2022-informational\n   :target: https://blog.jaraco.com/skeleton\n\n.. image:: https://tidelift.com/badges/package/pypi/zipp\n   :target: https://tidelift.com/subscription/pkg/pypi-zipp?utm_source=pypi-zipp&utm_medium=readme\n\n\nA pathlib-compatible Zipfile object wrapper. Official backport of the standard library\n`Path object <https://docs.python.org/3.8/library/zipfile.html#path-objects>`_.\n\n\nCompatibility\n=============\n\nNew features are introduced in this third-party library and later merged\ninto CPython. The following table indicates which versions of this library\nwere contributed to different versions in the standard library:\n\n.. list-table::\n   :header-rows: 1\n\n   * - zipp\n     - stdlib\n   * - 3.5\n     - 3.11\n   * - 3.3\n     - 3.9\n   * - 1.0\n     - 3.8\n\n\nUsage\n=====\n\nUse ``zipp.Path`` in place of ``zipfile.Path`` on any Python.\n\nFor Enterprise\n==============\n\nAvailable as part of the Tidelift Subscription.\n\nThis project and the maintainers of thousands of other packages are working with Tidelift to deliver one enterprise subscription that covers all of the open source you use.\n\n`Learn more <https://tidelift.com/subscription/pkg/pypi-zipp?utm_source=pypi-zipp&utm_medium=referral&utm_campaign=github>`_.\n\nSecurity Contact\n================\n\nTo report a security vulnerability, please use the\n`Tidelift security contact <https://tidelift.com/security>`_.\nTidelift will coordinate the fix and disclosure.\n', '_charset': None, 'preamble': None, 'epilogue': None, 'defects': [], '_default_type': 'text/plain'}

>>> i_metadata._headers
[('Metadata-Version', '2.1'), ('Name', 'zipp'), ('Version', '3.10.0'), ('Summary', 'Backport of pathlib-compatible object wrapper for zip files'), ('Home-page', 'https://github.com/jaraco/zipp'), ('Author', 'Jason R. Coombs'), ('Author-email', 'jaraco@jaraco.com'), ('Classifier', 'Development Status :: 5 - Production/Stable'), ('Classifier', 'Intended Audience :: Developers'), ('Classifier', 'License :: OSI Approved :: MIT License'), ('Classifier', 'Programming Language :: Python :: 3'), ('Classifier', 'Programming Language :: Python :: 3 :: Only'), ('Requires-Python', '>=3.7'), ('License-File', 'LICENSE'), ('Provides-Extra', 'docs'), ('Requires-Dist', "sphinx (>=3.5) ; extra == 'docs'"), ('Requires-Dist', "jaraco.packaging (>=9) ; extra == 'docs'"), ('Requires-Dist', "rst.linker (>=1.9) ; extra == 'docs'"), ('Requires-Dist', "furo ; extra == 'docs'"), ('Requires-Dist', "jaraco.tidelift (>=1.4) ; extra == 'docs'"), ('Provides-Extra', 'testing'), ('Requires-Dist', "pytest (>=6) ; extra == 'testing'"), ('Requires-Dist', "pytest-checkdocs (>=2.4) ; extra == 'testing'"), ('Requires-Dist', "pytest-flake8 ; extra == 'testing'"), ('Requires-Dist', "flake8 (<5) ; extra == 'testing'"), ('Requires-Dist', "pytest-cov ; extra == 'testing'"), ('Requires-Dist', "pytest-enabler (>=1.3) ; extra == 'testing'"), ('Requires-Dist', "jaraco.itertools ; extra == 'testing'"), ('Requires-Dist', "func-timeout ; extra == 'testing'"), ('Requires-Dist', "jaraco.functools ; extra == 'testing'"), ('Requires-Dist', "more-itertools ; extra == 'testing'"), ('Requires-Dist', 'pytest-black (>=0.3.7) ; (platform_python_implementation != "PyPy") and extra == \'testing\''), ('Requires-Dist', 'pytest-mypy (>=0.9.1) ; (platform_python_implementation != "PyPy") and extra == \'testing\'')]

>>> for (key, value) in i_metadata._headers:
    print(key, value)
    
Metadata-Version 2.1
Name zipp
Version 3.10.0
Summary Backport of pathlib-compatible object wrapper for zip files
Home-page https://github.com/jaraco/zipp
Author Jason R. Coombs
Author-email jaraco@jaraco.com
Classifier Development Status :: 5 - Production/Stable
Classifier Intended Audience :: Developers
Classifier License :: OSI Approved :: MIT License
Classifier Programming Language :: Python :: 3
Classifier Programming Language :: Python :: 3 :: Only
Requires-Python >=3.7
License-File LICENSE
Provides-Extra docs
Requires-Dist sphinx (>=3.5) ; extra == 'docs'
Requires-Dist jaraco.packaging (>=9) ; extra == 'docs'
Requires-Dist rst.linker (>=1.9) ; extra == 'docs'
Requires-Dist furo ; extra == 'docs'
Requires-Dist jaraco.tidelift (>=1.4) ; extra == 'docs'
Provides-Extra testing
Requires-Dist pytest (>=6) ; extra == 'testing'
Requires-Dist pytest-checkdocs (>=2.4) ; extra == 'testing'
Requires-Dist pytest-flake8 ; extra == 'testing'
Requires-Dist flake8 (<5) ; extra == 'testing'
Requires-Dist pytest-cov ; extra == 'testing'
Requires-Dist pytest-enabler (>=1.3) ; extra == 'testing'
Requires-Dist jaraco.itertools ; extra == 'testing'
Requires-Dist func-timeout ; extra == 'testing'
Requires-Dist jaraco.functools ; extra == 'testing'
Requires-Dist more-itertools ; extra == 'testing'
Requires-Dist pytest-black (>=0.3.7) ; (platform_python_implementation != "PyPy") and extra == 'testing'
Requires-Dist pytest-mypy (>=0.9.1) ; (platform_python_implementation != "PyPy") and extra == 'testing'
```

So `classifiers = [value for (key, value) in i_metadata._headers if key =='Classifier']` creates a list of all second tuple entries in `i_metadata._headers` where the first entry in the tuple equals 'Classifier'.
In the above example:
```
>>> classifiers = [classifier for (type, classifier) in i_metadata._headers if type =='Classifier']
>>> classifiers
['Development Status :: 5 - Production/Stable', 'Intended Audience :: Developers', 'License :: OSI Approved :: MIT License', 'Programming Language :: Python :: 3', 'Programming Language :: Python :: 3 :: Only']
```
This list is than used to extract the license informations from the classifiers.

I also noticed that the original code in python 3.10.8 also did not work as expected. I get the following output for `i_metadata`
```
>>> i_metadata['Classifier'] # upper case
'Development Status :: 5 - Production/Stable'
>>> i_metadata['classifier'] # lower case
'Development Status :: 5 - Production/Stable'
>>> i_metadata.json['classifier'] # lower case
['Development Status :: 5 - Production/Stable', 'Intended Audience :: Developers', 'License :: OSI Approved :: MIT License', 'Programming Language :: Python :: 3', 'Programming Language :: Python :: 3 :: Only']
>>> i_metadata.json['Classifier'] # upper case 
KeyError: 'Classifier'
```
So it should be at least (for python 3.10.8)
```
if 'classifier' in i_metadata.json:
                for classifier in i_metadata.json['classifier']:
```


The list generated by `[classifier for (type, classifier) in i_metadata._headers if type =='Classifier']` equals the one in ` i_metadata.json['classifier']`

Comment (a1lu) [reply cyclonedx_py/parser/environment.py:82]: Good catch, using python 3.7.15 I discovered that the code for python3.8 works in python3.7 too.

Comment (a1lu) [reply cyclonedx_py/parser/environment.py:82]: Since python 3.7 is the last supported version, is it enough to add support for it or do we need another branch for other (untested) versions?

**Comment by a1lu** [comment]
> Could you add proper (unit or integration) tests to prove that your changes actually work? The tests will be running on CI with every supported python version on every operating system.

Sure, Unfortunately I have only access to the github web IDE right now. I can do this probably during this week.

File: tests/test_parser_environment.py [hunk 2]
```diff
@@ -39,7 +39,9 @@ def test_simple(self) -> None:
         self.assertIsNotNone(c_tox)
         self.assertNotEqual(c_tox.purl.to_string(), c_tox.bom_ref.value)
         self.assertIsNotNone(c_tox.licenses)
-        self.assertEqual('MIT', c_tox.licenses.pop().license.name)
+        self.assertEqual(len(c_tox.licenses), 2)
+        self.assertEqual({LicenseChoice(license_=License(license_name="MIT License")),
```
[the remark this answers is not in the document]
Comment (a1lu) [reply tests/test_parser_environment.py:45]: Fixed

[the remark this answers is not in the document]
Comment (a1lu) [reply tests/test_parser_environment.py:45]: Fixed

**Comment by a1lu** [comment]
I added tests to check if both license entries are present. One from the license field and one from the classifiers.

Also I changed the classifier generation, as I discovered that the python3.10 version did not work on poetry run tox. No all versions use the same generator and the tests pass. Though I need to give mypy a hint to not generate an error.

I am going to squash the commits later (I am still only on github web IDE) and sign the commit.

File: cyclonedx_py/parser/environment.py [hunk 3]
```diff
@@ -75,18 +75,18 @@ def __init__(self, use_purl_bom_ref: bool = False) -> None:
...
-                    # Full list: https://pypi.python.org/pypi?%3Aaction=list_classifiers
-                    if str(classifier).startswith('License :: OSI Approved :: '):
-                        c.licenses.add(LicenseChoice(license_=License(
-                            license_name=str(classifier).replace('License :: OSI Approved :: ', '').strip()
-                        )))
-                    elif str(classifier).startswith('License :: '):
-                        c.licenses.add(LicenseChoice(license_=License(
-                            license_name=str(classifier).replace('License :: ', '').strip()
-                        )))
+            classifiers = [value for (key, value) in i_metadata._headers if key == 'Classifier']  # type: ignore
```
Comment (jkowalleck) [inline cyclonedx_py/parser/environment.py:78]: :x: `._headers` is a non-public property.
therefore, it should not be used. Please find the official/public getter for it.

**Review by jkowalleck — CHANGES_REQUESTED** [review]

File: tests/test_parser_environment.py [hunk 4]
```diff
@@ -39,7 +41,9 @@ def test_simple(self) -> None:
         self.assertIsNotNone(c_tox)
         self.assertNotEqual(c_tox.purl.to_string(), c_tox.bom_ref.value)
         self.assertIsNotNone(c_tox.licenses)
-        self.assertEqual('MIT', c_tox.licenses.pop().license.name)
```
[the remark this answers is not in the document]
Comment (a1lu) [reply tests/test_parser_environment.py:42]: The set contains both values. It gets `MIT` from the `License` tag and `MIT License` from the classifier.

File: cyclonedx_py/parser/environment.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (a1lu) [reply cyclonedx_py/parser/environment.py]: Solved. I found and use `i_metadata.get_all("Classifier")`.

Comment (a1lu) [reply cyclonedx_py/parser/environment.py]: Solved. I found and use `i_metadata.get_all("Classifier")`.

**Comment by jkowalleck** [comment]
Thanks for the code, @a1lu 

could you sign your request according to https://github.com/CycloneDX/cyclonedx-python/pull/431/checks?check_run_id=9242986125 
to signal that you are providing your code under the projects licenses?

**Comment by jkowalleck** [comment]
@madpah  please review this PR :D

**Review by jkowalleck — APPROVED** [review]

**Comment by a1lu** [comment]
@jkowalleck I squashed my commits and signed them.

Do we also want to parse the license file metadata entry? Or use the `get` function for `"Author"` and `"License"` too?  I could try to implement this in new PR.

**Comment by jkowalleck** [comment]
> @jkowalleck I squashed my commits and signed them.
great :+1: 



> Do we also want to parse the license file metadata entry? Or use the `get` function for `"Author"` and `"License"` too?
I would suggest to open an individual issue for each one of these topics, so we could discuss the need, the details and everything.

**Review by jkowalleck — APPROVED** [review]
